Fix errors when using clang on Windows

MaartenBent opened the pull request:
Fix some errors and support clang in CMake.

Also enable precompiled headers for samples. Note that cotire only enables precompiled headers when there are at least two source files.

Currently cotire does not correctly detect clang on Windows. I created [a PR](https://github.com/sakra/cotire/pull/161) to fix this. I did not include those changes in this PR, but will wait for the next cotire release. For now, apply the patch yourself or disable `wxBUILD_PRECOMP`.

When cotire is enabled, it seems to have the same issue with scintilla as XCode on macOS, so disable precompiled headers there.

Fixes error:
```
D:\dev\wxWidgets\src\msw\mediactrl_am.cpp:1580:10: error: case value evaluates to 4294966687, which cannot be narrowed to type 'DISPID' (aka 'long') [-Wc++11-narrowing]
    case 0xfffffd9f: // readystatechange in IActiveMovie2 and IMediaPlayer
         ^
```

Fixes error:
```
C:\msys64\mingw64\bin\ld: C:\msys64\mingw64\x86_64-w64-mingw32\lib/libuuid.a(lib64_libuuid_a-uuid.o):C:/building/msys64/mingw64/x86_64-w64-mingw32/include/shobjidl.h:9771: multiple definition of `IID_IShellItem'; ../lib/clang_x64_lib/wxmsw31u_core.a(dirdlg.cpp.obj):(.rdata[IID_IShellItem]+0x0): first defined here
```

Works around error:
```
Building CXX precompiled header libs/stc/cotire/wxscintilla_CXX_prefix.hxx.pch
In file included from D:/dev/build/wxWidgets/clang/libs/stc/cotire/wxscintilla_CXX_prefix.hxx:4:
In file included from D:\dev\build\wxWidgets\clang\libs\stc\cotire\wxscintilla_CXX_prefix.cxx:12:
In file included from C:\msys64\mingw64\include\c++\8.2.0\string:40:
In file included from C:\msys64\mingw64\include\c++\8.2.0\bits/char_traits.h:501:
C:\msys64\mingw64\include\c++\8.2.0\cstdint:48:11: error: no member named 'int8_t' in the global namespace
  using ::int8_t;
        ~~^
C:\msys64\mingw64\include\c++\8.2.0\cstdint:49:11: error: no member named 'int16_t' in the global namespace
  using ::int16_t;
        ~~^
C:\msys64\mingw64\include\c++\8.2.0\cstdint:50:11: error: no member named 'int32_t' in the global namespace
  using ::int32_t;
        ~~^
...
```

MaartenBent committed:
Fix c++11-narrowing error when using clang on Windows

Case value 0xfffffd9f results in the following error:
error: case value evaluates to 4294966687, which cannot be narrowed to type
'DISPID' (aka 'long') [-Wc++11-narrowing]

MaartenBent committed:
Fix multiple definition error when using clang on Windows

MaartenBent committed:
Support clang compiler on Windows when using CMake

MaartenBent committed:
Enable precompiled headers for samples

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -71,15 +71,11 @@ struct IShellItem : public IUnknown
-// Define this GUID in any case, even when __IShellItem_INTERFACE_DEFINED__ is
-// defined in the headers we might still not have it in the actual uuid.lib,
-// this happens with at least VC7 used with its original (i.e. not updated) SDK
-// and there is no harm in defining the GUID unconditionally.
```
Hmm, so it looks like this would break VC7 build? Not that I care much about it, and we should probably just officially drop support for it, but as long as we keep it, we probably shouldn't do this...

MaartenBent replied:
Somehow I assumed it wasn't supported in wx3.2. I'll revert it and add a `__VISUALC__` guard.

vadz commented on src/msw/mediactrl_am.cpp:
```diff
@@ -1530,7 +1530,7 @@ void wxAMMediaBackend::Move(int WXUNUSED(x), int WXUNUSED(y),
 void wxAMMediaEvtHandler::OnActiveX(wxActiveXEvent& event)
 {
-    switch(event.GetDispatchId())
+    switch ((unsigned long)event.GetDispatchId())
```
Sorry for nitpicking, but why not use more greppable `static_cast<>` instead? I'd also add a comment explaining why do we do this.

MaartenBent replied:
I did try `wxStaticCast`, but when that failed I just used c-style cast. I'll update it.

vadz replied:
FWIW `wxStaticCast` is to `static_cast<>` what `wxDynamicCast` to `dynamic_cast<>`, i.e. confusingly named similar but not at all identical macro.

vadz reviewed (commented):
Thanks for doing this!

MaartenBent committed:
Fix typos and whitespace in CMake files.

MaartenBent committed:
CMake: Improve precompiled headers for wxscintilla

The auto-generated header causes undefined members and identifiers in the
standard c++ headers when using clang on macOS or Windows.
Do not disable precompiled headers entirely but use the main Scintilla
header as prefix header so there is at least a small speedup.

MaartenBent committed:
Fix building taskbarbutton sample with precompiled headers
